File: php-implementation/examples/basic-usage.php
<?php

/**
 * Basic usage examples for PostgreSQL Advanced Features
 * 
 * Run: php examples/basic-usage.php
 */

require_once __DIR__ . '/../vendor/autoload.php';

use App\JsonbOperations;
use App\ConnectionPooling;

// Configuration
$config = [
    'db_host' => getenv('DB_HOST') ?: 'localhost',
    'db_port' => (int)(getenv('DB_PORT') ?: 5433),
    'db_database' => getenv('DB_DATABASE') ?: 'advanced_pg',
    'db_username' => getenv('DB_USERNAME') ?: 'pguser',
    'db_password' => getenv('DB_PASSWORD') ?: 'pgpass',
    'pgbouncer_host' => getenv('PGBOUNCER_HOST') ?: 'localhost',
    'pgbouncer_port' => (int)(getenv('PGBOUNCER_PORT') ?: 6433),
];

echo "=== PostgreSQL Advanced Features - Basic Usage ===\n\n";

// Initialize connection pooling
echo "1. Setting up connection pooling...\n";
$pooling = new ConnectionPooling($config);
$pdo = $pooling->getPooledConnection();
echo "✓ Connected via PgBouncer\n\n";

// Initialize JSONB operations
$jsonb = new JsonbOperations($pdo);

// Example 1: Query by brand
echo "2. Querying products by brand (Dell)...\n";
$dellProducts = $jsonb->getProductsByBrand('Dell');
foreach ($dellProducts as $product) {
    echo "  - {$product['name']} (Brand: {$product['brand']})\n";
}
echo "\n";

// Example 2: Query by CPU
echo "3. Querying products by CPU (i7)...\n";
$i7Products = $jsonb->getProductsByCpu('i7');
foreach ($i7Products as $product) {
    echo "  - {$product['name']} (CPU: {$product['cpu']})\n";
}
echo "\n";

// Example 3: Query by tag
echo "4. Querying products by tag (electronics)...\n";
$electronics = $jsonb->getProductsByTag('electronics');
echo "  Found " . count($electronics) . " products\n\n";

// Example 4: Complex search
echo "5. Complex search (Apple products, $500-$1000)...\n";
$results = $jsonb->searchProducts([
    'brand' => 'Apple',
    'min_price' => 500,
    'max_price' => 1000
]);
foreach ($results as $product) {
    $metadata = json_decode($product['metadata'], true);
    echo "  - {$product['name']}: \${$metadata['price']}\n";
}
echo "\n";

// Example 5: Update product price
echo "6. Updating product price...\n";
$firstProduct = $dellProducts[0] ?? null;
if ($firstProduct) {
    $oldMetadata = json_decode($firstProduct['metadata'], true);
    $oldPrice = $oldMetadata['price'];
    $newPrice = $oldPrice * 0.9; // 10% discount
    
    $jsonb->updateProductPrice($firstProduct['id'], $newPrice);
    echo "  ✓ Updated {$firstProduct['name']} from \${$oldPrice} to \${$newPrice}\n\n";
    
    // Restore original price
    $jsonb->updateProductPrice($firstProduct['id'], $oldPrice);
}

// Example 6: Add and remove discount
echo "7. Adding discount to product...\n";
if ($firstProduct) {
    $jsonb->addProductDiscount($firstProduct['id'], 15);
    echo "  ✓ Added 15% discount\n";
    
    $jsonb->removeProductDiscount($firstProduct['id']);
    echo "  ✓ Removed discount\n\n";
}

// Example 7: Benchmark connection pooling
echo "8. Benchmarking connection pooling...\n";
$benchmark = $pooling->benchmarkConnections(50);
echo "  Iterations: {$benchmark['iterations']}\n";
echo "  Direct time: {$benchmark['direct_time']}s\n";
echo "  Pooled time: {$benchmark['pooled_time']}s\n";
echo "  Improvement: {$benchmark['improvement']}\n\n";

// Example 8: Simulate high load
echo "9. Simulating high load (50 concurrent queries)...\n";
$loadTest = $pooling->simulateHighLoad(50);
echo "  Queries executed: {$loadTest['queries_executed']}\n";
echo "  Total time: {$loadTest['total_time']}s\n";
echo "  Queries per second: {$loadTest['queries_per_second']}\n\n";

// Example 9: Get pool statistics
echo "10. PgBouncer pool statistics...\n";
$stats = $pooling->getPoolStats();
if (isset($stats['error'])) {
    echo "  Note: {$stats['error']}\n";
} else {
    foreach ($stats as $pool) {
        echo "  Database: {$pool['database']}\n";
        echo "  Active connections: {$pool['cl_active']}\n";
        echo "  Waiting clients: {$pool['cl_waiting']}\n";
        break; // Show first pool only
    }
}
echo "\n";

echo "=== All examples completed successfully! ===\n";
echo "\nNext steps:\n";
echo "  - Review the code in src/JsonbOperations.php\n";
echo "  - Try the exercises in exercises/\n";
echo "  - Read the guides in docs/\n";
echo "  - Experiment with your own queries\n";

File: ruby-implementation/examples/basic_usage.rb
#!/usr/bin/env ruby

# Basic usage examples for PostgreSQL Advanced Features
# 
# Run: ruby examples/basic_usage.rb

require 'sequel'
require_relative '../lib/jsonb_operations'
require_relative '../lib/connection_pooling'

# Configuration
direct_url = ENV['DATABASE_URL'] || 'postgresql://pguser:pgpass@localhost:5434/advanced_pg'
pooled_url = ENV['PGBOUNCER_URL'] || 'postgresql://pguser:pgpass@localhost:6434/advanced_pg'

puts "=== PostgreSQL Advanced Features - Basic Usage ===\n\n"

# Initialize connections
puts "1. Setting up connection pooling..."
pooling = ConnectionPooling.new(direct_url, pooled_url)
jsonb = JsonbOperations.new(pooling.pooled_db)
puts "✓ Connected via PgBouncer\n\n"

# Example 1: Query by brand
puts "2. Querying products by brand (Dell)..."
dell_products = jsonb.get_products_by_brand('Dell')
dell_products.each do |product|
  metadata = JSON.parse(product[:metadata])
  puts "  - #{product[:name]} (Brand: #{metadata['brand']})"
end
puts "\n"

# Example 2: Query by CPU
puts "3. Querying products by CPU (i7)..."
i7_products = jsonb.get_products_by_cpu('i7')
i7_products.each do |product|
  puts "  - #{product[:name]} (CPU: #{product[:cpu]})"
end
puts "\n"

# Example 3: Query by tag
puts "4. Querying products by tag (electronics)..."
electronics = jsonb.get_products_by_tag('electronics')
puts "  Found #{electronics.count} products\n\n"

# Example 4: Complex search
puts "5. Complex search (Apple products, $500-$1000)..."
results = jsonb.search_products(
  brand: 'Apple',
  min_price: 500,
  max_price: 1000
)
results.each do |product|
  metadata = JSON.parse(product[:metadata])
  puts "  - #{product[:name]}: $#{metadata['price']}"
end
puts "\n"

# Example 5: Update product price
puts "6. Updating product price..."
first_product = dell_products.first
if first_product
  metadata = JSON.parse(first_product[:metadata])
  old_price = metadata['price']
  new_price = (old_price * 0.9).round(2) # 10% discount
  
  jsonb.update_product_price(first_product[:id], new_price)
  puts "  ✓ Updated #{first_product[:name]} from $#{old_price} to $#{new_price}\n\n"
  
  # Restore original price
  jsonb.update_product_price(first_product[:id], old_price)
end

# Example 6: Add and remove discount
puts "7. Adding discount to product..."
if first_product
  jsonb.add_product_discount(first_product[:id], 15)
  puts "  ✓ Added 15% discount"
  
  jsonb.remove_product_discount(first_product[:id])
  puts "  ✓ Removed discount\n\n"
end

# Example 7: Get price statistics
puts "8. Getting price statistics..."
stats = jsonb.get_price_statistics
puts "  Total products: #{stats[:total_products]}"
puts "  Average price: $#{stats[:avg_price].to_f.round(2)}"
puts "  Min price: $#{stats[:min_price]}"
puts "  Max price: $#{stats[:max_price]}\n\n"

# Example 8: Get all tags
puts "9. Getting all unique tags..."
tags = jsonb.get_all_tags
puts "  Tags: #{tags.join(', ')}\n\n"

# Example 9: Benchmark connection pooling
puts "10. Benchmarking connection pooling..."
benchmark = pooling.benchmark_connections(50)
puts "  Iterations: #{benchmark[:iterations]}"
puts "  Direct time: #{benchmark[:direct_time]}s"
puts "  Pooled time: #{benchmark[:pooled_time]}s"
puts "  Improvement: #{benchmark[:improvement]}\n\n"

# Example 10: Simulate high load
puts "11. Simulating high load (50 concurrent queries)..."
load_test = pooling.simulate_high_load(50)
puts "  Queries executed: #{load_test[:queries_executed]}"
puts "  Total time: #{load_test[:total_time]}s"
puts "  Queries per second: #{load_test[:queries_per_second]}\n\n"

# Example 11: Get pool statistics
puts "12. Connection pool statistics..."
stats = pooling.get_pool_stats
puts "  Direct pool size: #{stats[:direct_pool][:size]}/#{stats[:direct_pool][:max_size]}"
puts "  Pooled pool size: #{stats[:pooled_pool][:size]}/#{stats[:pooled_pool][:max_size]}\n\n"

# Example 12: Test transaction pooling
puts "13. Testing transaction pooling..."
tx_test = pooling.test_transaction_pooling
puts "  Direct transaction time: #{tx_test[:direct_transaction_time]}s"
puts "  Pooled transaction time: #{tx_test[:pooled_transaction_time]}s"
puts "  Note: #{tx_test[:note]}\n\n"

# Cleanup
pooling.close_connections

puts "=== All examples completed successfully! ===\n"
puts "\nNext steps:"
puts "  - Review the code in lib/jsonb_operations.rb"
puts "  - Try the exercises in exercises/"
puts "  - Read the guides in docs/"
puts "  - Experiment with your own queries"
